x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ConfigurableClusterPrivileges.java

```diff
@@ -362,4 +395,252 @@ private interface Fields {
+            return builder.field(
+                Fields.MANAGE.getPreferredName(),
+                Map.of(Fields.INDICES.getPreferredName(), indexPermissionGroups.stream().map(indexPatternPrivilege -> {
+                    Map<String, String[]> sortedMap = new TreeMap<>();
+                    sortedMap.put(Fields.NAMES.getPreferredName(), indexPatternPrivilege.indexPatterns());
+                    sortedMap.put(Fields.PRIVILEGES.getPreferredName(), indexPatternPrivilege.privileges());
+                    return sortedMap;
+                }).toList())
+            );
+        }
+
+        public static ManageRolesPrivilege parse(XContentParser parser) throws IOException {
+            expectedToken(parser.currentToken(), parser, XContentParser.Token.FIELD_NAME);
+            expectFieldName(parser, Fields.MANAGE);
+            expectedToken(parser.nextToken(), parser, XContentParser.Token.START_OBJECT);
+            expectedToken(parser.nextToken(), parser, XContentParser.Token.FIELD_NAME);
+            expectFieldName(parser, Fields.INDICES);
+            expectedToken(parser.nextToken(), parser, XContentParser.Token.START_ARRAY);
+            List<ManageRolesIndexPermissionGroup> indexPrivileges = new ArrayList<>();
+            Map<String, String[]> parsedArraysByFieldName = new HashMap<>();
+
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                expectedToken(token, parser, XContentParser.Token.START_OBJECT);
+                expectedToken(parser.nextToken(), parser, XContentParser.Token.FIELD_NAME);
+                String currentFieldName = parser.currentName();
+                expectedToken(parser.nextToken(), parser, XContentParser.Token.START_ARRAY);
+                parsedArraysByFieldName.put(currentFieldName, XContentUtils.readStringArray(parser, false));
+                expectedToken(parser.nextToken(), parser, XContentParser.Token.FIELD_NAME);
+                currentFieldName = parser.currentName();
+                expectedToken(parser.nextToken(), parser, XContentParser.Token.START_ARRAY);
+                parsedArraysByFieldName.put(currentFieldName, XContentUtils.readStringArray(parser, false));
+                expectedToken(parser.nextToken(), parser, XContentParser.Token.END_OBJECT);
+                indexPrivileges.add(
+                    new ManageRolesIndexPermissionGroup(
+                        parsedArraysByFieldName.get(Fields.NAMES.getPreferredName()),
+                        parsedArraysByFieldName.get(Fields.PRIVILEGES.getPreferredName())
+                    )
+                );
+            }
+            expectedToken(parser.nextToken(), parser, XContentParser.Token.END_OBJECT);
+            return new ManageRolesPrivilege(indexPrivileges);
+        }
+
+        public record ManageRolesIndexPermissionGroup(String[] indexPatterns, String[] privileges) implements Writeable {
```

Comment:
I might have missed this, but I don't see that we validate index patterns (`names`). They could be `null` or `empty`, which would trigger some assertions in `IndicesPermission.Group` when we build it in `requestPredicateSupplier`. I think we should apply the same validation as we do for indices permission group and disallow it to be `empty` or `null`. Meaning, we should not accept this:

```json
{
    "global": {
        "role": {
            "manage": {
                "indices": [
                    {
                        "names": [ ],
                        "privileges":["read"]
                    }
                ]
            }
        }
    }
}
```

Reply:
Great catch! There is a null check in the parser but not empty list check. Just added that validation. Will work on some unit tests for that too.